libcudacxx/include/cuda/std/__iterator/counted_iterator.h

```diff
@@ -0,0 +1,456 @@
+  = default;
+#else // ^^^ !_CCCL_NO_CONCEPTS ^^^ / vvv _CCCL_NO_CONCEPTS vvv
+  _CCCL_TEMPLATE(class _I2 = _Iter)
+  _CCCL_REQUIRES(default_initializable<_I2>)
+  _LIBCUDACXX_HIDE_FROM_ABI constexpr counted_iterator() noexcept(is_nothrow_default_constructible_v<_I2>) {}
+#endif // _CCCL_NO_CONCEPTS
+
+  _LIBCUDACXX_HIDE_FROM_ABI constexpr counted_iterator(_Iter __iter, iter_difference_t<_Iter> __n) noexcept(
+    is_nothrow_move_constructible_v<_Iter>)
+      : __current_(_CUDA_VSTD::move(__iter))
+      , __count_(__n)
+  {
+    _CCCL_ASSERT(__n >= 0, "__n must not be negative.");
+  }
+
+  _CCCL_TEMPLATE(class _I2)
+  _CCCL_REQUIRES(convertible_to<const _I2&, _Iter>)
+  _LIBCUDACXX_HIDE_FROM_ABI constexpr counted_iterator(const counted_iterator<_I2>& __other) noexcept(
+    is_nothrow_constructible_v<_Iter, const _I2&>)
+      : __current_(__other.__current_)
+      , __count_(__other.__count_)
+  {}
+
+  _CCCL_TEMPLATE(class _I2)
+  _CCCL_REQUIRES(assignable_from<_Iter&, const _I2&>)
+  _LIBCUDACXX_HIDE_FROM_ABI constexpr counted_iterator&
+  operator=(const counted_iterator<_I2>& __other) noexcept(is_nothrow_assignable_v<_Iter&, const _I2&>)
+  {
+    __current_ = __other.__current_;
+    __count_   = __other.__count_;
+    return *this;
+  }
+
+  [[nodiscard]] _LIBCUDACXX_HIDE_FROM_ABI constexpr const _Iter& base() const& noexcept
+  {
+    return __current_;
+  }
+
+  [[nodiscard]] _LIBCUDACXX_HIDE_FROM_ABI constexpr _Iter base() &&
+  {
+    return _CUDA_VSTD::move(__current_);
+  }
+
+  [[nodiscard]] _LIBCUDACXX_HIDE_FROM_ABI constexpr iter_difference_t<_Iter> count() const noexcept
+  {
+    return __count_;
+  }
+
+  [[nodiscard]] _LIBCUDACXX_HIDE_FROM_ABI constexpr decltype(auto) operator*()
+  {
+    _CCCL_ASSERT(__count_ > 0, "Iterator is equal to or past end.");
+    return *__current_;
+  }
+
+  _CCCL_TEMPLATE(class _I2 = _Iter)
+  _CCCL_REQUIRES(__dereferenceable<const _I2>)
+  [[nodiscard]] _LIBCUDACXX_HIDE_FROM_ABI constexpr decltype(auto) operator*() const
+  {
+    _CCCL_ASSERT(__count_ > 0, "Iterator is equal to or past end.");
+    return *__current_;
+  }
+
+  _CCCL_TEMPLATE(class _I2 = _Iter)
+  _CCCL_REQUIRES(contiguous_iterator<_I2>)
+  [[nodiscard]] _LIBCUDACXX_HIDE_FROM_ABI constexpr auto operator->() const noexcept
+  {
+    return _CUDA_VSTD::to_address(__current_);
+  }
+
+  _LIBCUDACXX_HIDE_FROM_ABI constexpr counted_iterator& operator++()
+  {
+    _CCCL_ASSERT(__count_ > 0, "Iterator already at or past end.");
+    ++__current_;
+    --__count_;
+    return *this;
+  }
+
+  _CCCL_TEMPLATE(class _I2 = _Iter)
+  _CCCL_REQUIRES((!forward_iterator<_I2>) )
+  _LIBCUDACXX_HIDE_FROM_ABI decltype(auto) operator++(int)
+  {
+    _CCCL_ASSERT(__count_ > 0, "Iterator already at or past end.");
+    --__count_;
+    auto __guard = _CUDA_VSTD::__make_exception_guard([&] {
+      ++__count_;
+    });
+    __current_++;
+    __guard.__complete();
+    return _CUDA_VSTD::move(__current_);
```

Comment:
i'm following up with LWG, but a literal reading of the spec (https://eel.is/c++draft/counted.iter.nav#4) says that this function should return `void`. the specified _Effects:_ para has no `return`.